tensorflow/c/c_api.cc

```diff
@@ -2458,10 +2458,20 @@ void TF_AddGradientsWithPrefix(TF_Graph* g, const char* prefix, TF_Output* y,
 
     // Update g->name_map with the name_map from the scope, which will contain
     // the new gradient ops.
+
+    // Uses y as the pivot for control edges.
+    Node* pivot = &y->oper->node;
```

Comment:
Unconditionally adding control edges from the pivot can result in different behavior, since the gradient computation doesn't necessarily depend on the forward pass completing first. For instance, if y = x + 1, then dy/dx = 1, and the gradient computation doesn't depend on the forward pass at all. So adding the control edge could result in performance regressions, or even functional differences if there are stateful ops involved.

I'm not sure what the best approach is to avoid this regression... maybe we can assign a WhileContext to all nodes in a while loop, not just the Exit nodes, and only add the pivot edge if *y is in a while loop?

Reply:
I'm not very familiar with the Java API, but it wraps the [C API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.h), right? (This is the API that includes TF_AddGradients.) The C API is built on top of the core API, so I believe Exit nodes created via the C API should have WhileContexts, and although the C API doesn't expose the WhileContexts, the TF_AddGradients implementation has access to it. Does this sound right?

Reply:
You're absolutely correct that the Java API does wrap around c_api.cc, so you do have access to WhileContext from the C API. However we think the issue is rather that Exit nodes are given a WhileContext in the [BuildWhileLoop function](https://github.com/tensorflow/tensorflow/blob/9590c4c32dd4346ea5c35673336f5912c6072bf2/tensorflow/cc/ops/while_loop.cc#L240), which is part of the C++ library, but TF Java never calls upon/does not have access to the BuildWhileLoop function. It appears that there isn't any or much support for creation of while loops through a function call in TF Java; we have been creating while loops by directly creating primitive ops. Therefore the Exit nodes in TF Java that we have been using do not have a WhileContext. 

I see that there is some [while loop support in c_api.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.h#L1076), but these structs/functions are not exposed to the Java API through the JNI bridge. Is that intentional?

Reply:
Ah, didn't realize the Java API was creating while loops directly out of primivite ops... how does it create the backedge to close the loop? I thought going through TF_NewWhile/TF_FinishWhile was the only way to actually create a loop.

I'm not sure why they're not exposed through the JNI bridge, probably someone just needs to add them?

Reply:
We had exposed the graph.updateEdge function through the JNI: [PR (not accepted)](https://github.com/tensorflow/tensorflow/pull/24859), [Issue ](https://github.com/tensorflow/tensorflow/issues/24231)

We would definitely be interested in using TF_NewWhile/TF_FinishWhile if it can be exposed through the JNI. However, @asimshankar did suggest pursuing the functional while loop in TF Java like python's `while_v2` as per our discussion [here](https://github.com/tensorflow/tensorflow/issues/24231), which is why our PR for updateEdge was not accepted. Thoughts?

Reply:
Using the functional while loop would be great, especially if TF_NewWhile/TF_FinishWhile could be modified so all language bindings could take advantage of it. This would be more work than simply calling TF_NewWhile/TF_FinishWhile as-is, but would help avoid issues like the one this PR is trying to solve. Would you be interested in taking on this project? I would be happy to advise.

Where is the JNI defined? Is there a reason it couldn't be used to call TF_NewWhile/TF_FinishWhile?

Reply:
Unfortunately, we won't have the time to take on adding functional while loops to C API, but we would be more than happy to expose the TF_NewWhile/TF_FinishWhile functions through the [JNI](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/java/src/main/native). The changes would be similar to the updateEdge PR in that those changes will also simply expose a lower level function to TF Java. (We'd still need a solution for this addGradients issue though.) 

If it's not intentionally unexposed, then someone probably just didn't add them to the JNI :)

Reply:
Yeah I don't think it's intentional, everything in c_api.h is designed to be exposed to other language bindings.

So it sounds like our current plan of action is:
1. Expose TF_NewWhile + TF_FinishWhile + TF_WhileParams in JNI, and use those to construct while loops in TF Java.
2. Change C++ while_loop implementation to add WhileContext to all nodes in the while loop, not just Exits.
3. Use this to determine if we need special behavior to handle constants in a while_loop... maybe by adding the control dep in this PR, or maybe by adding an Enter node. I think adding the Enter node is better.

Does this sound right?

Reply:
Sounds great. And then eventually perhaps `TF_NewWhile` and friends can be used to produce functional while loops - but that would be a C implementation change only.

And you're right, the only reason `TF_NewWhile` and friends aren't exposed through JNI is that nobody had asked so far :) (In general, the TensorFlow maintainers have made it so that the other language APIs are suited for loading and executing models. We haven't had the time to flesh out full support for graph construction, but are open to community contributions for them). Thanks!

Reply:
Sounds great, we'll get on it. Thanks so much for your responses! We'll let you know if we have any questions/concerns in the meantime.

Reply:
Sounds good, thanks!

Reply:
Hi Skye! We were talking about #2 on the list (setting the WhileContext for all nodes created by the while loop). We were running into problems with setting the WhileContext for the nodes created by the body and cond subgraphs. Essentially, because we don't have a way to access the nodes after they are created, we figured that we would add WhileContext to scope (like control dependencies) so that it is added when the nodes are created. However, since the WhileContext needs the exit nodes, we would have to modify the WhileContext given to the body and cond nodes after the exit nodes are created. We noticed that there aren't setters in WhileContext and everything's a constant, which made us think that this might not be the right idea. We were wondering if there's a different intended design that we missed or if you have any insights.

We were also wondering if there is a better place to have the discussion since our comments don't have to do with this pull request anymore.

Reply:
Hi Skye and Asim,

We're also getting started on exposing TF_NewWhile etc. through the JNI as discussed, and wanted to follow up on how you'd prefer this functionality to be exposed.

We see two options for doing this:

1) We could create an all-in-one while loop method for Java, similar to Python's tf.while_loop. This way, TF_NewWhile and TF_FinishWhile would be called at the JNI level (this would also mean the TF_WhileParams would be generated and passed along at the JNI level). 

2) Alternatively, we could expose TF_NewWhile, TF_FinishWhile, and TF_AbortWhile individually, as in c_api. If we do this, we would need to return some while params object from the Java newWhile method, which the user would modify and pass to finishWhile or abortWhile. If you'd prefer this way, how would you suggest exposing TF_WhileParams in Java?

Reply:
We could move this discussion to developers@tensorflow.org (you can join [here](https://groups.google.com/a/tensorflow.org/forum/#!forum/developers) if you haven't already). @samdow wanna copy your question about WhileContexts into a new thread there? I need to look into a little, and it's possible others in the community have ideas.

@melissagrueter I think option 1 is preferable, assuming it's not too burdensome to pass in cond and body functions in Java.

Reply:
Thanks for your help! I've posted it [here](https://groups.google.com/a/tensorflow.org/forum/?utm_medium=email&utm_source=footer#!topic/developers/IjuGQNcD6Kw). Just wanted to leave it here so I don't lose it 😊